### PR TITLE
fix never_loop

### PR DESCRIPTION
**Pull request by camsteffen**
Fixes https://github.com/Manishearth/rust-clippy/issues/1586.

This is my first time contributing to rust-clippy or even working with rust internals! Let me know if you can find any counterexamples or have any other feedback.

The logic essentially works like this...

```
fn never_loop() -> bool {
    if the loop contains a ‘continue’ (this loop) {
        false
    } else if all branches contain ‘break’ or ‘return’ or ‘continue’ (outer loop) {
        true
    } else {
        false
    }
}
```

### CODE REVIEW

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -424,47 +422,100 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
...
         ExprTup(ref es) |
-        ExprMethodCall(_, _, ref es) => es.iter().any(|e| never_loop_expr(e)),
-        ExprCall(ref e, ref es) => never_loop_expr(e) || es.iter().any(|e| never_loop_expr(e)),
-        ExprBlock(ref block) => never_loop_block(block),
-        ExprStruct(_, _, ref base) => base.as_ref().map_or(false, |e| never_loop_expr(e)),
+        ExprMethodCall(_, _, ref es) => es.iter().any(|e| contains_continue_expr(e, dest)),
+        ExprCall(ref e, ref es) => contains_continue_expr(e, dest) || es.iter().any(|e| contains_continue_expr(e, dest)),
+        ExprBlock(ref block) => contains_continue_block(block, dest),
+        ExprStruct(_, _, ref base) => base.as_ref().map_or(false, |e| contains_continue_expr(e, dest)),
+        ExprAgain(d) => d.target_id.opt_id().map_or(false, |id| id == *dest),
```
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Shouldn't this be `map_or(true`?

Comment (camsteffen) [inline clippy_lints/src/loops.rs]: I'm not sure. `opt_id()` only returns `None` in cases where there is a [LoopIdError](https://manishearth.github.io/rust-internals-docs/rustc/hir/enum.LoopIdError.html) which means the `continue` is invalid. So I figured if the `continue` is invalid then I should ignore it.

**Comment by camsteffen** [comment]
I added a bunch of tests and found some cases that needed fixing. Also, I expanded the lint to cover all kinds of loops, not just `loop { ... }`.

**Comment by camsteffen** [comment]
Hmm. It looks like I flagged some never_loop instances elsewhere in the code after widening the scope of the lint. Do we like this? Should I go fix those?

**Comment by oli-obk** [comment]
Congratulations! You found a bug in clippy! 🎉

Yes, please fix those

File: clippy_lints/src/loops.rs (hunk 1 above)
Comment (camsteffen) [inline clippy_lints/src/loops.rs]: @oli-obk is that correct?

**Comment by mcarton** [comment]
I see there is a problem with `doc.rs`, don't bother to much fixing it because of #1799.

File: clippy_lints/src/loops.rs (hunk 1 above)
Comment (oli-obk) [inline clippy_lints/src/loops.rs]: Yea. I thought it was None when there was no label.

**Comment by oli-obk** [comment]
Restarting ci

**Comment by camsteffen** [comment]
I left the issue in `doc.rs`. Will that be fixed soon with #1799?

**Comment by oli-obk** [comment]
No clue. Feel free to allow the lint on the offending statement

**Comment by camsteffen** [comment]
I just fixed it. Should be ready to merge.

**Comment by oli-obk** [comment]
Yay! Thanks.

**Pull request merged**